README.md
# Bevillingsplatform for Ballerup (C-BUR)
[![pipeline status](https://git.magenta.dk/bevillingsplatform/bevillingsplatform/badges/develop/pipeline.svg)](https://git.magenta.dk/bevillingsplatform/bevillingsplatform/commits/develop)
[![coverage report](https://git.magenta.dk/bevillingsplatform/bevillingsplatform/badges/develop/coverage.svg)](https://git.magenta.dk/bevillingsplatform/bevillingsplatform/commits/develop)

## Installation

TL;DR: To get a running development environment run:

```bash
git@git.magenta.dk:bevillingsplatform/bevillingsplatform.git
cd bevillingsplatform
docker-compose up -d --build frontend
```

You can now reach the frontend at http://localhost:8080. The frontend will proxy
all request to the backend.

Run backend tests with:
```bash
docker-compose exec bev pytest
```


### Docker

The repository contains a `Dockerfile`. This is the recommended way to install
bevillingsplatform both as a developer and in production.

All releases are pushed to Docker Hub at
[`magentaaps/bevillingsplatform`](https://hub.docker.com/r/magentaaps/bevillingsplatform)
under the `latest` tag.

To run bevillingsplatform in docker you need a running docker daemon. To install
docker we refer you to the [official
documentation](https://docs.docker.com/install/).

To configure the django inside the image, add your setting to
`/code/settings.ini`. This is easiest done by binding a file from the host
machine to this file.

The container requires a connection to a postgres database server. It is
configured with the `DATABASE_*` settings. The database server must have a user
and a database object. It can be created with the help of
`/docker/postgres-initdb.d/20-create-db-and-user.sh`. This file can easily be
mounted into `/docker-entrypoint-initdb.d/` in [the official postgres docker
image](https://hub.docker.com/_/postgres).

You can start a the container with:
```bash
docker run -p 8000:8000 -v $PWD/dev-settings.ini:/code/settings.ini magentaaps/bevillingsplatform:latest
```

This will pull the image from Docker Hub and starts a container in the
foreground. The `-p 8000:8000` [binds
port](https://docs.docker.com/engine/reference/commandline/run/#publish-or-expose-port--p---expose)
`8000` of the host machine to port `8000` on the container. The `-v
$PWD/dev-settings.ini:/code/settings.ini`
[binds](https://docs.docker.com/engine/reference/commandline/run/#mount-volume--v---read-only)
the `dev-settings.ini` file into the container at the location where django will
pick it up.

If successful you should see the container migrating the database and finally
```
[2019-06-13 09:18:48 +0000] [1] [INFO] Listening at: http://0.0.0.0:8000 (1)
```

when the gunicorn server starts up. You should now be able to reach the server
from the host at ``http://localhost:8000``.


If you continue to see `01/30 Unable to connect to database.` your database
configuration most likely wrong. Remember if you set `DATABASE_HOST=localhost`
the container will try to connect to a database in the same container, not the
host machine.


#### Static files

The docker image will serve the static files for the Vue frontend, Django REST
framework and Django Admin from gunicorn both in development and in production.
Normally it is not good practice to serve static files from gunicorn for
security and performance reasons. We use
[whitenoise](https://pypi.org/project/whitenoise/) most of these conserns and
generally don't expect many users. If you still want to serve it from another
service, all the files are copied to `/static` on container startup. This can
easily be mounted to a webserver.


#### Logs

The gunicorn error log is output on `STDERR`. It can be inspected with `docker
logs`. The gunicorn access log is written to `/log/access.log`. The django log
is written to `/log/django-debug.log`.


#### User permissions

The `Dockerfile` creates and runs the application as the `bev` user.
This user will own all the files generated by the application. This user has a
``UID`` and ``GID`` of 72050.

If you want to use another ``UID/GID``, you can specify it as the
`--user=uid:gid` [overwrite
flag](https://docs.docker.com/engine/reference/run/#user) for the ``docker run``
command or [in
docker-compose](https://docs.docker.com/compose/compose-file/#domainname-hostname-ipc-mac_address-privileged-read_only-shm_size-stdin_open-tty-user-working_dir)
If you change the `UID/GID`, the `/log` and `/static` volumes may not have the
right permissions. It is recommended to only use
[bind](https://docs.docker.com/storage/bind-mounts/) if you overwrite the user
and set the same user as owner of the directory you bind.

If some process inside the container needs to write files to locations other
than `/static` or `/log`, you need to mount a volume with the right permissions.
An example is `./manage.py makemigrations` trying to write to
`/code/backend/core/migrations`. If you bind `/code` to your host system, make
sure that the user with UID 72050 have write permissions to
`backend/core/migrations`. This can be done with `chmod o+w migrations` on your
host where you grant all user permission to write.


#### Test

All the requirements for tests included in the docker image. You can run the
test from inside a container with `pytest`.

##### tox
`tox` is also installed, but it tries to create a virtual environments inside
the container. This is messy an will fail because the application user does not
have permission to write files. Don't use `tox` inside the container.


### Docker-compose

You can use ``docker-compose`` to start up bevillingsplatform and related
service such as postgres and postfix.

A `docker-compose.yml` for development is included. It includes the settings
to connect them. It starts four services:

- `frontend`: the vue frontend reachable at  http://localhost:8080,
- `bev`: the django backend,
- `db`: a [postgres database server](https://hub.docker.com/_/postgres)
- `postfix`: a [postfix email server](https://hub.docker.com/r/catatnight/postfix).

Normally the backend image also serves the frontend code, but to ease frontend
development, we include a frontend service that run [`vue-cli-service
serve`](https://cli.vuejs.org/guide/cli-service.html). The frontend proxies
requests to the backend. The exact list of proxied endpoints can be seen in
`frontend/vue.config.js`.

`docker-compose.yml` also mounts the current directory in the container and
automatically restarts the server on changes to the backend files. This enables
you to edit the backend files and the server will be reloaded automatically.

To pull the images and start the three service run:
```bash
docker-compose up -d --build frontend
```

The `-d` flag move the services to the background. You can inspect the output of
them with `docker-compose logs <name>` where `<name>` is the name of the service
in `docker-compose.yml`. The `--build` flag builds the newest docker image for
`bevillingsplatform` from the local `Dockerfile`.

To stop the service again run `docker-compose stop`. This will stop the
services, but the data will persist. To completely remove the containers and
data run `docker-compose down -v`.


#### Tests and shell access

To run the backend test, execute: `docker-compose exec bev ./manage.py test`. It
will connect to the running docker container and execute the tests.

To get shell access to the backend run `docker-compose exec bev bash`.

If you want to write files from inside the container, make sure the `bev` user
have permission to do so. See [User permissions](#user-permissions).


#### Postgres initialisation

The ``docker-compose.yml`` file contains a service named ``bev-cp``. Its purpose
is to copy the files needed to initialize the database and database user to a
volume. This volume can then be mounted to the postgres image to automatically
initialize the database. This functionality is not needed by default because the
needed files are mounted directly from the host. It is included as an example
when you want to use an environment closer to production.


## Other Readmes

[Backend README](backend/README.rst)

[Frontend README](frontend/README.rst)

## Licensing

Copyright (c) 2019 Magenta Aps

Bevillingsplatform is free software; you may use, study, modify and
distribute it under the terms of version 2.0 of the Mozilla Public
License. See the LICENSE file for details. If a copy of the MPL was not
distributed with this file, You can obtain one at
http://mozilla.org/MPL/2.0/.

All source code in this and the underlying directories is subject to
the terms of the Mozilla Public License, v. 2.0.

NEWS
* Version 2.3.0, 2020-01-09

New in this version:

Features:

* Modify the URL for the rate tabel (taksttabel) to a more general one.
* Nice-ify django admin for Payments and PaymentSchedules and allow search on payment id.
* Add pydocstyle compliance.
* Add sphinx docs generation

Bug fixes:

* Fix incorrect tests dependent on current year.

* Version 2.2.3, 2019-12-12

New in this version:

* Changes to PRISM file generation.
* Enforce rules for activities on grant.
* Disable edit for appropriation fields on granted activities.
* Enable date validation for activities.
* Add filtering on payment type.
* Fix payment CPR filtering.
* Small improvements to logging.
* Make tox work locally.
* Add frontend documentation.
* Add cronjobs for docker.
* Fix date filtering.
* Fix failing tests.
* Fix paths in settings.
* Update Django from 2.2.1 to 2.2.4

* Version 2.2.2, 2019-11-28

Hotfix:

* Fix broken migration.


* Version 2.2.1, 2019-11-25

New in this version:

Features:

* Mark payments for SD Løn along with fictive ones.

Bug fixes:

* Fix hover text.
* Display of Indsatstrappen fixed.
* Fix ordering of Indsatstrappen.
* Recipient info stayed in GUI even though payment method was changed to
  "internal".
* Empty "not found" text when displaying "Mine sager".


* Version 2.2.0, 2019-11-21

New in this version:

Features:

* It is now possible to find payments from a payment ID.
* Case worker can now be changed on several cases in one action.
* A log of all pending and sent emails is now kept and accessible in the
  Django admin interface.
* Payment ID and account string is displayed in the Django admin
  interface.
* Generally improved interface for searching and displaying cases.
* Fictive payments are clearly marked as fictive in payment plans.
* Fictive payments are marked as paid in the database on the day they
  are due.
* Field added in API to indicate whether a payment can be paid
  manually or not.
* Payments that are paid as Salary (through SD-Løn) or cash or are
  fictive may not be edited manually.
* Payments are paginated to avoid too long loading times.
* Payments are now sorted by payment date.
* Payments are sorted by *ascending* payment date.
* Indsatstrappen is now a classification to be maintained in the Django
  Admin interface.
* Section (of the law, from the appropriation) is added to the payment
  emails.
* Emails are sent for all approved payments, for all combinations of
  payment and recipient types.
* Complex logic for generation of account string.
* Payment dialog improved.
* Integration to KMD PRISME accounting system.
* Information about citizen included in display of appropriation.


Bug fixes:

* Don't throw an exception if users attempt to access the API without
  logging in, just deny access.
* If more than one user profile is sent from SAML IdP, don't crash -
  choose the *highest* one.
* "Mixed content error" on some pages (on internal test server).
* Many small and big improvements to styling and usability.
* Function deciding if case is expired also looked at DELETED
  activities.


* Version 2.0.1, 2019-11-11

New in this version:

* Add support for Service Provider certificates through PySaml2.


* Version 2.0.0, 2019-11-06

New in this version:

* Implement SAML SSO login.
* Implement user rights levels.
* Add preliminary Prism file generation.
* Implement GUI for editing payments.
* Add support for "fictive" payments.
* Add support for negative and zero payments.
* Add support for paid amounts and paid date for payments.
* Update payment summation to include paid amounts when able.
* Add new payment ID for payment plans.
* Add account strings for payments.
* Add API filtering for several endpoints.
* Remove the "udbetaling til firma" payment option.
* Fix a bug when creating an activity.
* Fix redirect when setting a payment paid.
* Add missing verbose names in Django admin.

* Version 1.1.1, 2019-10-30

Hotfix release.

New in this version:

* Deleted main activity no longer blocks for creating a new main activity.
* Granted activities are now explicitly included in the appropration PDF.
* Fix activities still being checked for granting when closing the grant dialog.
* Fix not being able to grant an expected main activity.
* Fix invalid XML in OS2forms.xml.
* Add missing constraint for creating supplementary activities based on allowed main activities.

* Version 1.1.0, 2019-10-04

New in this version:

* Fixed approval button when there's nothing to approve.
* Fixed missing activities from appropriation PDF.
* Fix spelling error in logout message.
* For payment to a person with SDLøn, tax card is mandatory.
* Use user first_name and last_name instead of initials for Sagsbehandler dropdown.
* Fix stop dates on supplementary activities.
* Fix link to rates document.
* Correctly calculate the expected amount for expected activities.
* Correct forms for modifying effort steps (Indsatstrappe) in Djang Admin.
* Clear frontend errors correctly.
* Rearrange autologin scripts in frontend.
* Change recommended browser text.
* Suppress not writeable warning from ipython.


* Version 1.0.0, 2019-09-27

First production release. New in this version:

* KLE number and SBYS template info moved from Section to new
  SectionInfo class in the ManyToMany relation.
* Activities are granted individually, not all at once for each
  appropriation.
* Missing logo fixed/supplied.
* Various GUI and UX improvements.
* Prevent expected changes from starting in the past.
* Make user supply day of month for monthly payments - handle month end
  correctly.
* Browser compatibility fixes.
* Fix missing update of family relations.
* Improved handling of backend error messages.
* New API fields for expected and granted totals for activities.
* Appropriation PDF nicified and adapted to the new approval scheme.
* SBSYS integration (os2forms.xml) fixed.
* Cases *must* have a team, this field is now non-nullable.
* Activities with status EXPECTED are now soft-deleted.

backend/core/utils.py
# Copyright (C) 2019 Magenta ApS, http://magenta.dk.
# Contact: info@magenta.dk.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
"""Utilities used by other parts of this app."""


import os
import logging
import requests
import datetime
import itertools

from dateutil.relativedelta import relativedelta

from django.template.loader import get_template
from django.core.mail import EmailMessage

from django.conf import settings
from django.template.loader import render_to_string
from django.core.mail import send_mail
from django.utils.translation import gettext_lazy as _
from django.utils.html import strip_tags
from django.db import transaction

from constance import config

from weasyprint import HTML
from weasyprint.fonts import FontConfiguration

from service_person_stamdata_udvidet import get_citizen

from core import models


logger = logging.getLogger(__name__)


def get_next_interval(from_date, payment_frequency):
    """Calculate the next date based on start date and payment frequency."""
    from core.models import PaymentSchedule

    if payment_frequency == PaymentSchedule.DAILY:
        new_start = from_date + relativedelta(days=1)
    elif payment_frequency == PaymentSchedule.WEEKLY:
        new_start = from_date + relativedelta(weeks=1)
    elif payment_frequency == PaymentSchedule.BIWEEKLY:
        new_start = from_date + relativedelta(weeks=2)
    elif payment_frequency == PaymentSchedule.MONTHLY:
        new_start = from_date.replace(day=1) + relativedelta(months=1)
    else:
        raise ValueError(_("ukendt betalingsfrekvens"))
    return new_start


def get_person_info(cpr):
    """Get CPR data on a person and his/her relations."""
    if settings.USE_SERVICEPLATFORM:
        func = get_cpr_data
    else:
        func = get_cpr_data_mock
    result = func(cpr)
    if not result:
        return None
    for relation in result["relationer"]:
        relation_cpr = relation["cprnr"]
        relation_data = func(relation_cpr)
        if relation_data:
            relation.update(relation_data)
    return result


def get_cpr_data(cpr):
    """Get CPR data from Serviceplatformen."""
    if not os.path.isfile(settings.SERVICEPLATFORM_CERTIFICATE_PATH):
        logger.info(
            "serviceplatform certificate path: %s is not a file",
            settings.SERVICEPLATFORM_CERTIFICATE_PATH,
        )
        return None
    try:
        result = get_citizen(
            service_uuids=settings.SERVICEPLATFORM_UUIDS,
            certificate=settings.SERVICEPLATFORM_CERTIFICATE_PATH,
            cprnr=cpr,
            production=settings.USE_SERVICEPLATFORM_PROD,
        )
        return result
    except requests.exceptions.HTTPError:
        logger.exception("get_cpr_data requests error")
        return None


def get_cpr_data_mock(cpr):
    """Use test data in place of the real 'get_cpr_data' for develop/test."""
    result = {
        "statsborgerskab": "5100",
        "efternavn": "Jensen",
        "postdistrikt": "Næstved",
        "foedselsregistreringssted": "Myndighedsnavn for landekode: 5902",
        "boernUnder18": "false",
        "civilstandsdato": "1991-03-21+01:00",
        "adresseringsnavn": "Jens Jensner Jensen",
        "fornavn": "Jens Jensner",
        "tilflytningsdato": "2001-12-01+01:00",
        "markedsfoeringsbeskyttelse": "true",
        "vejkode": "1759",
        "standardadresse": "Sterkelsvej 17 A,2",
        "etage": "02",
        "koen": "M",
        "status": "80",
        "foedselsdato": "1978-04-27+01:00",
        "vejnavn": "Sterkelsvej",
        "statsborgerskabdato": "1991-09-23+02:00",
        "adressebeskyttelse": "false",
        "stilling": "Sygepl ske",
        "gaeldendePersonnummer": "2704785263",
        "vejadresseringsnavn": "Sterkelsvej",
        "civilstand": "G",
        "alder": "59",
        "relationer": [
            {"cprnr": "0123456780", "relation": "aegtefaelle"},
            {"cprnr": "1123456789", "relation": "barn"},
            {"cprnr": "2123456789", "relation": "barn"},
            {"cprnr": "3123456789", "relation": "barn"},
            {"cprnr": "0000000000", "relation": "mor"},
            {"cprnr": "0000000000", "relation": "far"},
        ],
        "postnummer": "4700",
        "husnummer": "017A",
        "vejviserbeskyttelse": "true",
        "kommunekode": "370",
    }
    return result


def send_activity_email(subject, template, activity):
    """Send an email concerning an updated activity."""
    html_message = render_to_string(template, {"activity": activity})
    send_mail(
        subject,
        strip_tags(html_message),
        config.DEFAULT_FROM_EMAIL,
        [config.TO_EMAIL_FOR_PAYMENTS],
        html_message=html_message,
    )


def send_activity_created_email(activity):
    """Send an email because an activity was created."""
    subject = _("Aktivitet oprettet")
    template = "emails/activity_created.html"
    send_activity_email(subject, template, activity)


def send_activity_updated_email(activity):
    """Send an email because an activity was updated."""
    subject = _("Aktivitet opdateret")
    template = "emails/activity_updated.html"
    send_activity_email(subject, template, activity)


def send_activity_expired_email(activity):
    """Send an email because an activity has expired."""
    subject = _("Aktivitet udgået")
    template = "emails/activity_expired.html"
    send_activity_email(subject, template, activity)


def send_appropriation(appropriation, included_activities=None):
    """Generate PDF and XML files from appropriation and send them to SBSYS.

    :param appropriation: the Appropriation from which to generate PDF and XML.
    :param included_activities: Activities which should be explicitly included.

    """
    if included_activities is None:
        included_activities_qs = models.Activity.objects.none()
    else:
        # Convert to queryset.
        included_activities_qs = models.Activity.objects.filter(
            id__in=(a.id for a in included_activities)
        )

    today = datetime.date.today()
    approved_main_activities = (
        appropriation.activities.filter(
            activity_type=models.MAIN_ACTIVITY, status=models.STATUS_GRANTED
        )
        .exclude(end_date__lt=today)
        .union(
            included_activities_qs.filter(activity_type=models.MAIN_ACTIVITY)
        )
    )

    approved_suppl_activities = (
        appropriation.activities.filter(
            activity_type=models.SUPPL_ACTIVITY, status=models.STATUS_GRANTED
        )
        .exclude(end_date__lt=today)
        .union(
            included_activities_qs.filter(activity_type=models.SUPPL_ACTIVITY)
        )
    )

    render_context = {
        "appropriation": appropriation,
        "main_activities": approved_main_activities,
        "supplementary_activities": approved_suppl_activities,
    }

    # Get SBSYS template and KLE number from main activity.
    section_info = appropriation.section_info
    render_context["kle_number"] = section_info.kle_number
    render_context["sbsys_template_id"] = section_info.sbsys_template_id

    # Generate os2forms.xml
    xml_template = get_template(settings.SBSYS_XML_TEMPLATE)
    xml_data = xml_template.render(context=render_context)
    xml_file_name = "os2forms.xml"

    # Generate PDF
    html_template = get_template(settings.SBSYS_APPROPRIATION_TEMPLATE)
    html_data = html_template.render(context=render_context)

    # Configure fonts for correct rendering.
    font_config = FontConfiguration()
    pdf_data = HTML(string=html_data).write_pdf(font_config=font_config)
    pdf_file_name = f"{appropriation.sbsys_id}.pdf"

    # Send as email
    msg = EmailMessage()
    msg.subject = "Bevillingsskrivelse"
    msg.body = ""
    msg.from_email = config.DEFAULT_FROM_EMAIL
    msg.to = [config.SBSYS_EMAIL]
    msg.attachments = [
        (xml_file_name, xml_data, "text/xml"),
        (pdf_file_name, pdf_data, "application/pdf"),
    ]

    msg.send()


def saml_before_login(user_data):  # noqa: D401
    """Hook called after userdata is received from IdP, before login."""
    user_changed = False
    username = user_data["username"][0]
    user = models.User.objects.get(username=username)
    if "team" in user_data:
        # SAML data comes as lists with one element.
        team_name = user_data["team"][0]
        # This is safe, user exists.
        team, _ = models.Team.objects.get_or_create(
            name=team_name, defaults={"leader": user}
        )
        if team != user.team:
            user.team = team
            user_changed = True
    if "bos_profile" in user_data:
        profile = models.User.max_profile(user_data["bos_profile"])
        if profile != user.profile:
            user.profile = profile
            is_admin = profile == models.User.ADMIN
            # Admin status is controlled by these flags.
            user.is_staff = user.is_superuser = is_admin
            user_changed = True
    if user_changed:
        user.save()


def saml_create_user(user_data):  # noqa: D401
    """Hook called after user is created in DB, before login."""
    username = user_data["username"][0]
    user = models.User.objects.get(username=username)
    if "team" in user_data:
        # SAML data comes as lists with one element.
        team_name = user_data["team"][0]
    else:
        team_name = config.DEFAULT_TEAM_NAME

    # This is safe, user exists.
    team, _ = models.Team.objects.get_or_create(
        name=team_name, defaults={"leader": user}
    )
    user.team = team

    if "bos_profile" in user_data:
        profile = models.User.max_profile(user_data["bos_profile"])
        user.profile = profile
        is_admin = profile == models.User.ADMIN
        # Admin status is controlled by these flags.
        user.is_staff = user.is_superuser = is_admin
    user.save()


# Economy integration releated stuff - for the time being, only PRISM.
# TODO: At some point, factor out customer specific third party integrations.


def format_prism_financial_record(payment, line_no, record_no):
    """Format a single financial record for PRISM, on a single line.

    This follows documentation provided by Ballerup Kommune based on
    KMD's interface specification GQ311001Q for financial records (transaction
    type G69).
    """
    # The fields that are hard coded *never* change.
    # We specify them as variables below, but in reality we might as
    # well hardcode them in the actual output.

    reg_location = "000"
    interface_type = "G69"
    org_type = "01"
    post_type = "NOR"
    line_format = "FLYD"

    # Line number is given as 5 chars with leading zeroes, org unit as 4 chars
    # with leading zeroes, as per the specification.
    header = (
        f"{reg_location}{interface_type}{line_no:05d}"
        + f"{config.PRISM_ORG_UNIT:04d}{org_type}{post_type}{line_format}"
    )

    # Now the actual posting fields. These are marked with a leading '&' and
    # must come in increasing order by field number.

    """
    103 - machine number. Here, always '00482'. Number, 5 chars with leading
    zeroes.

    104 - "ekspeditionsløbenr". Enumerating each posting, 1, 2, ... Number, 7
    chars with leading zeroes. The "record_no" parameter.

    110 - posting date, format 'YYYYMMDD'

    111 - account number. Account string for this payment. Number, 10 chars.

    112 - amount. Number, 12 chars with leading zeroes + 1 trailing char for
    the sign ('+' or ' ' for plus, '-' for minus - the latter is not relevant
    here).

    113 - debit or credit; 'D' for debit, 'K' for credit. Will always be 'D'.

    114 - fiscal year. Use the fiscal year of the date parameter.

    117 - udbetalingshenvisningsnummer - date + machine number + expedition
    number.

    132 - recipient number code - always '02' for CPR number.

    133 - beneficiary - 10 digits, CPR number.

    153 - posting text.
    """

    case_cpr = payment.payment_schedule.activity.appropriation.case.cpr_number
    fields = {
        "103": f"{config.PRISM_MACHINE_NO:05d}",
        "104": f"{record_no:07d}",
        "110": f"{payment.date.strftime('%Y%m%d')}",
        "111": f"{payment.account_string}",
        "112": f"{int(payment.amount*100):012d} ",
        "113": "D",
        "114": f"{payment.date.year}",
        "132": "02",
        "133": f"{case_cpr}",
        "153": f"{payment.payment_schedule.payment_id}",
    }
    fields["117"] = fields["110"] + fields["103"] + fields["104"]

    field_string = "".join(
        f"&{field_no}{fields[field_no]}" for field_no in sorted(fields)
    )
    # Record id header followed by field string.
    return header + field_string


def format_prism_payment_record(payment, line_no, record_no):
    """Format a single payment record for PRISM, on a single line.

    This follows documentation provided by Ballerup Kommune based on
    KMD's interface specification GF200001Q for creditor records
    (transaction type G68).
    """
    # First, we format the header.
    # The header has the following fields that never change and might as
    # well be hard coded:

    reg_location = "000"
    interface_type = "G68"
    transaction_type = "01"
    line_format = "1"

    header = (
        f"{reg_location}{interface_type}{line_no:05d}"
        + f"{config.PRISM_ORG_UNIT:04d}{transaction_type}{line_format}"
    )

    # Now the mandatory fields. In the file, they are preceded with "&"
    # and must come in non-decreasing order.
    """
    02 - org unit, municipality number - 4 digits with leading zeroes.

    03 - organisation type, always given as '00'. 2 digits.

    08 - amount. 11 digits with leading zeroes.

    09 - sign. '+' for larger than zero, '-' for less. Always '+'.

    10 - ident. code. '02' for CPR number and payment to person.

    11 - payee. CPR number - 10 digits.

    12 - date. 8 digits, 'YYYYMMDD'.

    16 - posteringshenvisningsnummer. As 117 in the finance records -
    date + machine number + record number - 8 chars + 5 chars + 7 chars.

    17 - Payment ID. 20 digits with leading zeroes.

    40 - posting text.
    """

    payment_id = payment.payment_schedule.payment_id
    fields = {
        "02": f"{config.PRISM_ORG_UNIT:04d}",
        "03": "00",
        "08": f"{int(payment.amount*100):011d}",
        "09": "+",
        "10": "02",
        "11": f"{payment.recipient_id}",
        "12": f"{payment.date.strftime('%Y%m%d')}",
        "17": f"{payment_id:020d}",
        "40": f"Fra Ballerup Kommune ref: {payment_id}",
    }
    fields[
        "16"
    ] = f"{fields['12']}{config.PRISM_MACHINE_NO:05d}{record_no:07d}"

    field_string = "".join(
        f"&{field_no}{fields[field_no]}" for field_no in sorted(fields)
    )
    # Record id header followed by field string.
    return header + field_string


def due_payments_for_prism(date):
    """Return payments which are due on date and should be sent to PRISM."""
    return models.Payment.objects.filter(
        date=date,
        recipient_type=models.PaymentSchedule.PERSON,
        payment_method=models.CASH,
        paid=False,
        payment_schedule__activity__status=models.STATUS_GRANTED,
        payment_schedule__fictive=False,
        amount__gt=0,
    )


def generate_records_for_prism(due_payments):
    """Generate the list of records for writing to PRISM file."""
    prism_records = (
        (
            format_prism_financial_record(p, line_no=2 * i - 1, record_no=i),
            format_prism_payment_record(p, line_no=2 * i, record_no=i),
        )
        for i, p in enumerate(due_payments, 1)
    )
    # Flatten for easier handling.
    prism_records = list(itertools.chain(*prism_records))

    return prism_records


@transaction.atomic
def export_prism_payments_for_date(date=None):
    """Process payments and output a file for PRISME."""
    # The output directory is not configurable - this is mapped through Docker.
    output_dir = settings.PRISM_OUTPUT_DIR
    # Date = tomorrow if not given. We need "tomorrow" to set payment date.
    tomorrow = datetime.datetime.now() + relativedelta(day=1)
    if not date:
        date = tomorrow

    # The microseconds are included to avoid accidentally overwriting
    # tomorrow's file.
    filename = f"{output_dir}/{date.strftime('%Y%m%d')}_{tomorrow.microsecond}"
    payments = due_payments_for_prism(date)
    if not payments.exists():
        # No payments
        return
    with open(filename, "w") as f:
        # Generate and write preamble.

        """
        The fields given below never change and might as well be hard coded in
        the output:
        """

        hdisp = " "  # Blank, must be there.
        media_type = "6"  # Don't ask.
        evolbr = "      "  # 6 blanks - once again, don't ask.
        mixed = "1"
        trans_code = "Z300"  # Identifies transaction start.

        user_number = f"{config.PRISM_ORG_UNIT:04d}"  # Org unit.
        day_of_year = tomorrow.timetuple().tm_yday  # Day of year.

        preamble_string = (
            f"{trans_code}{hdisp}{user_number}{media_type}{evolbr}"
            + f"{day_of_year}{mixed}"
        )
        f.write(f"{preamble_string}\n")
        # Generate and write the records.
        prism_records = generate_records_for_prism(payments)
        f.write("\n".join(prism_records))

        # Generate and write the final line.
        cslutd = "SLUTD"  # Don't ask.
        fantrec = f"{len(prism_records):05d}"
        f.write(f"\n{cslutd}{fantrec}\n")
    # Register all payments as paid.
    for p in payments:
        p.paid = True
        p.paid_amount = p.amount
        p.paid_date = tomorrow
        p.save()
    # Return filename for info and verification.
    return filename
